support/k8s/k8s-workload-registrar/mode-crd/webhook/webhook_svid.go

```diff
@@ -0,0 +1,180 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	svidv1 "github.com/spiffe/spire-api-sdk/proto/spire/api/server/svid/v1"
+	spiretypes "github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+)
+
+const (
+	certDirMode   = os.FileMode(0o600)
+	certsFileMode = os.FileMode(0o644)
+	keyFileMode   = os.FileMode(0o600)
+	certsFileName = "tls.crt"
+	keyFileName   = "tls.key"
+)
+
+type SVIDConfig struct {
+	Cluster            string
+	Log                logrus.FieldLogger
+	Namespace          string
+	S                  svidv1.SVIDClient
+	TrustDomain        string
+	WebhookCertDir     string
+	WebhookServiceName string
+}
+
+type SVID struct {
+	c         SVIDConfig
+	id        spiffeid.ID
+	beginsAt  time.Time
+	expiresAt time.Time
+}
+
+// NewSVID sets up the data structure and mints a new SVID for the webhok
+func NewSVID(ctx context.Context, config SVIDConfig) (*SVID, error) {
+	err := os.MkdirAll(config.WebhookCertDir, certDirMode)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := spiffeid.New(config.TrustDomain, "k8s-workload-registrar", config.Cluster, "webhook")
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate Spiffe ID: %w", err)
+	}
+
+	return &SVID{
+		c:  config,
+		id: id,
+	}, nil
+}
+
+// MintSVID requests the SPIRE Server to mint a new SVID for the webhook
+func (e *SVID) MintSVID(ctx context.Context, key crypto.Signer) (err error) {
+	e.c.Log.Debug("Minting new SVID for webhook")
+	// Generate key if not passed in
+	if key == nil {
+		key, err = ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		if err != nil {
+			return fmt.Errorf("unable to generate Key: %w", err)
+		}
+	}
+
+	// Generate Certificate Signing Request
+	csr, err := x509.CreateCertificateRequest(rand.Reader, &x509.CertificateRequest{
+		URIs: []*url.URL{e.id.URL()},
+		DNSNames: []string{
+			e.c.WebhookServiceName + "." + e.c.Namespace + ".svc",
+		},
+	}, key)
+	if err != nil {
+		return fmt.Errorf("unable to generate CSR: %w", err)
+	}
+
+	// Mint new SVID
+	resp, err := e.c.S.MintX509SVID(ctx, &svidv1.MintX509SVIDRequest{
+		Csr: csr,
+	})
+	if err != nil {
+		return fmt.Errorf("unable to make Mint SVID Request: %w", err)
+	}
+	e.beginsAt = time.Now()
```

Comment:
This is probably good enough for this purpose but in general this timestamp should be obtained from the certificate. It shouldn't be a problem here, but I thought I'd point it out as a red flag.

Reply:
yes, youre right. Do you know offhand if the leaf certificate is the first or the last in the chain in the mint svid response?

Reply:
leaf is first

Reply:
this change is in.